File: packages/serverCore/ts/src/secretStore/envelope.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// CSS1 envelope encryption for the Caracal Secret Store: per-secret data keys wrapped by the master KEK.

import { createCipheriv, createDecipheriv, createHash, randomBytes } from 'node:crypto'

const KEY_BYTES = 32
const NONCE_BYTES = 12
const TAG_BYTES = 16
const MAGIC = Buffer.from('CSS1', 'ascii')
const KEK_ID_BYTES = 8
const DEK_BLOCK_BYTES = NONCE_BYTES + KEY_BYTES + TAG_BYTES
const MIN_ENVELOPE_BYTES = MAGIC.length + KEK_ID_BYTES + DEK_BLOCK_BYTES + NONCE_BYTES + TAG_BYTES
const DEK_AAD = Buffer.from('caracal.css1.dek', 'utf8')

// Envelope layout, all lengths fixed except the value block:
//   magic(4) | kekId(8) | dekNonce(12) | dekCt(48) | valNonce(12) | valCt(n+16)
// The data key is random per envelope and sealed under the KEK; the value is sealed
// under the data key with the caller's AAD binding the ciphertext to its logical
// location, so a blob moved to another row or table refuses to decrypt.

export function kekId(kek: Buffer): Buffer {
  return createHash('sha256').update(kek).digest().subarray(0, KEK_ID_BYTES)
}

function aeadSeal(key: Buffer, nonce: Buffer, plaintext: Buffer, aad: Buffer): Buffer {
  const cipher = createCipheriv('chacha20-poly1305', key, nonce, { authTagLength: TAG_BYTES })
  cipher.setAAD(aad, { plaintextLength: plaintext.length })
  const enc = Buffer.concat([cipher.update(plaintext), cipher.final()])
  return Buffer.concat([enc, cipher.getAuthTag()])
}

function aeadOpen(key: Buffer, nonce: Buffer, sealed: Buffer, aad: Buffer): Buffer {
  const tag = sealed.subarray(sealed.length - TAG_BYTES)
  const body = sealed.subarray(0, sealed.length - TAG_BYTES)
  const decipher = createDecipheriv('chacha20-poly1305', key, nonce, { authTagLength: TAG_BYTES })
  decipher.setAAD(aad, { plaintextLength: body.length })
  decipher.setAuthTag(tag)
  return Buffer.concat([decipher.update(body), decipher.final()])
}

export function sealEnvelope(kek: Buffer, plaintext: Buffer, aad: string): Buffer {
  if (kek.length !== KEY_BYTES) throw new Error(`kek must be ${KEY_BYTES} bytes`)
  const dek = randomBytes(KEY_BYTES)
  try {
    const dekNonce = randomBytes(NONCE_BYTES)
    const valNonce = randomBytes(NONCE_BYTES)
    const dekCt = aeadSeal(kek, dekNonce, dek, DEK_AAD)
    const valCt = aeadSeal(dek, valNonce, plaintext, Buffer.from(aad, 'utf8'))
    return Buffer.concat([MAGIC, kekId(kek), dekNonce, dekCt, valNonce, valCt])
  } finally {
    dek.fill(0)
  }
}

export function openEnvelope(kek: Buffer, envelope: Buffer, aad: string): Buffer {
  if (kek.length !== KEY_BYTES) throw new Error(`kek must be ${KEY_BYTES} bytes`)
  if (envelope.length < MIN_ENVELOPE_BYTES) throw new Error('secret envelope too short')
  if (!envelope.subarray(0, MAGIC.length).equals(MAGIC)) throw new Error('secret envelope has unknown format')
  let offset = MAGIC.length
  const id = envelope.subarray(offset, offset + KEK_ID_BYTES)
  offset += KEK_ID_BYTES
  if (!id.equals(kekId(kek))) throw new Error('secret envelope was sealed under a different KEK')
  const dekNonce = envelope.subarray(offset, offset + NONCE_BYTES)
  offset += NONCE_BYTES
  const dekCt = envelope.subarray(offset, offset + KEY_BYTES + TAG_BYTES)
  offset += KEY_BYTES + TAG_BYTES
  const valNonce = envelope.subarray(offset, offset + NONCE_BYTES)
  offset += NONCE_BYTES
  const valCt = envelope.subarray(offset)
  const dek = aeadOpen(kek, dekNonce, dekCt, DEK_AAD)
  try {
    return aeadOpen(dek, valNonce, valCt, Buffer.from(aad, 'utf8'))
  } finally {
    dek.fill(0)
  }
}

function weakKekReason(key: Buffer): string {
  let allSame = true
  let ascending = true
  let descending = true
  let alternating = true
  for (let i = 1; i < key.length; i++) {
    if (key[i] !== key[0]) allSame = false
    if (key[i] !== key[i - 1]! + 1) ascending = false
    if (key[i] !== key[i - 1]! - 1) descending = false
    if (i >= 2 && key[i] !== key[i % 2]) alternating = false
  }
  if (allSame && key[0] === 0) return 'must not be all zeros'
  if (allSame) return 'must not repeat the same byte'
  if (ascending || descending) return 'must not use a sequential byte pattern'
  if (alternating) return 'must not use a repeating byte pattern'
  return ''
}

function parseKek(name: string, raw: string): Buffer {
  const key = Buffer.from(raw, 'hex')
  if (key.length !== KEY_BYTES) {
    throw new Error(`${name} must be ${KEY_BYTES} bytes, got ${key.length}`)
  }
  const reason = weakKekReason(key)
  if (reason) throw new Error(`${name} ${reason}`)
  return key
}

// The Secret Store master key: 32 bytes of hex delivered by file-backed environment,
// never stored in the database, so a database compromise alone cannot decrypt secrets.
export function loadSecretStoreKek(): Buffer {
  const raw = process.env.SECRET_STORE_KEK
  if (!raw) throw new Error('SECRET_STORE_KEK is required')
  return parseKek('SECRET_STORE_KEK', raw)
}

// The active keyring, current key first. SECRET_STORE_KEK_PREVIOUS keeps the
// retiring key readable during a rotation window while envelopes are re-sealed.
export function loadSecretStoreKeks(): Buffer[] {
  const keys = [loadSecretStoreKek()]
  const previous = process.env.SECRET_STORE_KEK_PREVIOUS
  if (previous) keys.push(parseKek('SECRET_STORE_KEK_PREVIOUS', previous))
  return keys
}

// The key identifier embedded in an envelope, for rotation tooling that routes
// or skips rows without decrypting them.
export function envelopeKekId(envelope: Buffer): Buffer {
  if (envelope.length < MIN_ENVELOPE_BYTES) throw new Error('secret envelope too short')
  if (!envelope.subarray(0, MAGIC.length).equals(MAGIC)) throw new Error('secret envelope has unknown format')
  return envelope.subarray(MAGIC.length, MAGIC.length + KEK_ID_BYTES)
}

// The operational seal path: always the current key.
export function sealSecretEnvelope(plaintext: Buffer, aad: string): Buffer {
  return sealEnvelope(loadSecretStoreKek(), plaintext, aad)
}

// The operational open path: the embedded kekId routes each envelope to the key
// that sealed it, so reads keep working across a KEK rotation window.
export function openSecretEnvelope(envelope: Buffer, aad: string): Buffer {
  const id = envelopeKekId(envelope)
  for (const kek of loadSecretStoreKeks()) {
    if (id.equals(kekId(kek))) return openEnvelope(kek, envelope, aad)
  }
  throw new Error('secret envelope was sealed under a different KEK')
}

File: packages/core/go/secretstore/envelope.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// CSS1 envelope encryption for the Caracal Secret Store: per-secret data keys wrapped by the master KEK.

package secretstore

import (
	"bytes"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"math"
	"os"

	"golang.org/x/crypto/chacha20poly1305"
)

const (
	keyBytes    = 32
	nonceBytes  = 12
	tagBytes    = 16
	kekIDBytes  = 8
	dekBlockLen = nonceBytes + keyBytes + tagBytes
)

var (
	magic  = []byte("CSS1")
	dekAAD = []byte("caracal.css1.dek")
)

func minEnvelopeLen() int {
	return len(magic) + kekIDBytes + dekBlockLen + nonceBytes + tagBytes
}

// KEKID identifies the master key that wrapped an envelope's data key.
func KEKID(kek []byte) []byte {
	sum := sha256.Sum256(kek)
	return sum[:kekIDBytes]
}

// Seal produces a CSS1 envelope:
//
//	magic(4) | kekId(8) | dekNonce(12) | dekCt(48) | valNonce(12) | valCt(n+16)
//
// The data key is random per envelope and sealed under the KEK; the value is sealed
// under the data key with aad binding the ciphertext to its logical location, so a
// blob moved to another row or table refuses to decrypt.
func Seal(kek, plaintext []byte, aad string) ([]byte, error) {
	if len(kek) != keyBytes {
		return nil, fmt.Errorf("kek must be %d bytes", keyBytes)
	}
	if len(plaintext) > math.MaxInt-minEnvelopeLen() {
		return nil, errors.New("plaintext too large to seal")
	}
	dek := make([]byte, keyBytes)
	if _, err := rand.Read(dek); err != nil {
		return nil, err
	}
	defer clear(dek)
	kekAEAD, err := chacha20poly1305.New(kek)
	if err != nil {
		return nil, err
	}
	dekAEAD, err := chacha20poly1305.New(dek)
	if err != nil {
		return nil, err
	}
	dekNonce := make([]byte, nonceBytes)
	valNonce := make([]byte, nonceBytes)
	if _, err := rand.Read(dekNonce); err != nil {
		return nil, err
	}
	if _, err := rand.Read(valNonce); err != nil {
		return nil, err
	}
	envelope := make([]byte, 0, minEnvelopeLen()+len(plaintext))
	envelope = append(envelope, magic...)
	envelope = append(envelope, KEKID(kek)...)
	envelope = append(envelope, dekNonce...)
	envelope = kekAEAD.Seal(envelope, dekNonce, dek, dekAAD)
	envelope = append(envelope, valNonce...)
	return dekAEAD.Seal(envelope, valNonce, plaintext, []byte(aad)), nil
}

// Open decrypts a CSS1 envelope produced by Seal in any Caracal language runtime.
func Open(kek, envelope []byte, aad string) ([]byte, error) {
	if len(kek) != keyBytes {
		return nil, fmt.Errorf("kek must be %d bytes", keyBytes)
	}
	if len(envelope) < minEnvelopeLen() {
		return nil, errors.New("secret envelope too short")
	}
	offset := 0
	if !bytes.Equal(envelope[:len(magic)], magic) {
		return nil, errors.New("secret envelope has unknown format")
	}
	offset += len(magic)
	if !bytes.Equal(envelope[offset:offset+kekIDBytes], KEKID(kek)) {
		return nil, errors.New("secret envelope was sealed under a different KEK")
	}
	offset += kekIDBytes
	dekNonce := envelope[offset : offset+nonceBytes]
	offset += nonceBytes
	dekCt := envelope[offset : offset+keyBytes+tagBytes]
	offset += keyBytes + tagBytes
	valNonce := envelope[offset : offset+nonceBytes]
	offset += nonceBytes
	kekAEAD, err := chacha20poly1305.New(kek)
	if err != nil {
		return nil, err
	}
	dek, err := kekAEAD.Open(nil, dekNonce, dekCt, dekAAD)
	if err != nil {
		return nil, errors.New("secret envelope data key rejected")
	}
	defer clear(dek)
	dekAEAD, err := chacha20poly1305.New(dek)
	if err != nil {
		return nil, err
	}
	plaintext, err := dekAEAD.Open(nil, valNonce, envelope[offset:], []byte(aad))
	if err != nil {
		return nil, errors.New("secret envelope value rejected")
	}
	return plaintext, nil
}

func weakKEKReason(b []byte) string {
	allSame := true
	ascending := true
	descending := true
	alternating := true
	for i := 1; i < len(b); i++ {
		if b[i] != b[0] {
			allSame = false
		}
		if int(b[i]) != int(b[i-1])+1 {
			ascending = false
		}
		if int(b[i]) != int(b[i-1])-1 {
			descending = false
		}
		if i >= 2 && b[i] != b[i%2] {
			alternating = false
		}
	}
	switch {
	case allSame && b[0] == 0:
		return "must not be all zeros"
	case allSame:
		return "must not repeat the same byte"
	case ascending || descending:
		return "must not use a sequential byte pattern"
	case alternating:
		return "must not use a repeating byte pattern"
	default:
		return ""
	}
}

func loadKEK(name string) ([]byte, error) {
	raw := os.Getenv(name)
	if raw == "" {
		return nil, fmt.Errorf("%s is required", name)
	}
	b, err := hex.DecodeString(raw)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", name, err)
	}
	if len(b) != keyBytes {
		return nil, fmt.Errorf("%s must be %d bytes, got %d", name, keyBytes, len(b))
	}
	if reason := weakKEKReason(b); reason != "" {
		return nil, fmt.Errorf("%s %s", name, reason)
	}
	return b, nil
}

// Keyring holds the active Secret Store master keys, current first. Seal always
// uses the current key; Open routes each envelope by its embedded kekId so reads
// keep working across a KEK rotation window.
type Keyring struct {
	keys [][]byte
	ids  [][]byte
}

func NewKeyring(keys ...[]byte) (*Keyring, error) {
	if len(keys) == 0 {
		return nil, errors.New("keyring requires at least one key")
	}
	ring := &Keyring{}
	for _, key := range keys {
		if len(key) != keyBytes {
			return nil, fmt.Errorf("kek must be %d bytes", keyBytes)
		}
		ring.keys = append(ring.keys, key)
		ring.ids = append(ring.ids, KEKID(key))
	}
	return ring, nil
}

// LoadKeyring reads the master keys: 32 bytes of hex delivered by file-backed
// environment, never stored in the database, so a database compromise alone cannot
// decrypt secrets. SECRET_STORE_KEK_PREVIOUS keeps the retiring key readable while
// envelopes are re-sealed during a rotation.
func LoadKeyring() (*Keyring, error) {
	current, err := loadKEK("SECRET_STORE_KEK")
	if err != nil {
		return nil, err
	}
	keys := [][]byte{current}
	if os.Getenv("SECRET_STORE_KEK_PREVIOUS") != "" {
		previous, err := loadKEK("SECRET_STORE_KEK_PREVIOUS")
		if err != nil {
			return nil, err
		}
		keys = append(keys, previous)
	}
	return NewKeyring(keys...)
}

func (k *Keyring) Seal(plaintext []byte, aad string) ([]byte, error) {
	return Seal(k.keys[0], plaintext, aad)
}

func (k *Keyring) Open(envelope []byte, aad string) ([]byte, error) {
	if len(envelope) < minEnvelopeLen() {
		return nil, errors.New("secret envelope too short")
	}
	if !bytes.Equal(envelope[:len(magic)], magic) {
		return nil, errors.New("secret envelope has unknown format")
	}
	id := envelope[len(magic) : len(magic)+kekIDBytes]
	for i, keyID := range k.ids {
		if bytes.Equal(id, keyID) {
			return Open(k.keys[i], envelope, aad)
		}
	}
	return nil, errors.New("secret envelope was sealed under a different KEK")
}
